syntax = "proto3";

import "google/protobuf/empty.proto";
import "google/protobuf/struct.proto";

package aegis_proto;

// --------------------------------- Marker --------------------------------- //

service Marker {
    rpc Start (google.protobuf.Empty) returns (google.protobuf.Empty) {}
    rpc Stop (google.protobuf.Empty) returns (google.protobuf.Empty) {}
}

// --------------------------------- Recorder -------------------------------- //

service Recorder {
    rpc Listen (google.protobuf.Empty) returns (stream RecorderListenResponse) {}
}

message RecorderListenResponse {
    string command = 1;
}

// -------------------------------- Searcher --------------------------------- //

service Searcher {
    rpc Tree (OptionalObjectRequest) returns (TreeResponse) {}
    rpc ListenTreeChanges (google.protobuf.Empty) returns (stream TreeChangeResponse) {}
    rpc Find (FindRequest) returns (FindResponse) {}
}

message OptionalObjectRequest {
    optional string object = 1;
}

message TreeResponse {
    repeated TreeNode nodes = 1;
}

message TreeNode {
    string object = 1;
    repeated TreeNode nodes = 2;
}

message TreeChangeResponse {
    oneof change_type {
        ObjectAdded added = 1;
        ObjectRemoved removed = 2;
        ObjectReplaced replaced = 3;
    }
}

message ObjectAdded {
    string parent = 1;
    string object = 2;
}

message ObjectRemoved {
    string parent = 1;
    string object = 2;
}

message ObjectReplaced {
    string from = 1;
    string to = 2;
}

message FindRequest {
    string object = 1;
}

message FindResponse {
    repeated string objects = 1;
}

// --------------------------------- Object ---------------------------------- //

service Object {
    rpc Parent (ObjectRequest) returns (ParentResponse) {}
    rpc Children (ObjectRequest) returns (ChildrenResponse) {}

    rpc InvokeMethod (InvokeMethodRequest) returns (google.protobuf.Empty) {}
    rpc SetProperty (SetPropertyRequest) returns (google.protobuf.Empty) {}

    rpc DumpMethods (ObjectRequest) returns (DumpMethodsResponse) {}
    rpc DumpProperties (ObjectRequest) returns (DumpPropertiesResponse) {}
}

message ObjectRequest {
    string object = 1;
}

message ParentResponse {
    optional string parent = 1;
}

message ChildrenResponse {
    repeated string children = 1;
}

message InvokeMethodRequest {
    string object = 1;
    string method = 2;
    repeated google.protobuf.Value arguments = 3;
}

message SetPropertyRequest {
    string object = 1;
    string property = 2;
    google.protobuf.Value value = 3;
}

message DumpMethodsResponse {
    repeated string methods = 1;
}

message DumpPropertiesResponse {
    repeated PropertyValue properties = 1;
}

message PropertyValue {
    string property = 1;
    google.protobuf.Value value = 2;
}